pkg/datapath/linux/route/reconciler/manager.go

```diff
@@ -0,0 +1,254 @@
+	}
+}
+
+func (m *DesiredRouteManager) FinalizeInitializer(initializer Initializer) {
+	if initializer.initialized != nil {
+		txn := m.db.WriteTxn(m.tbl)
+		defer txn.Commit()
+		initializer.initialized(txn)
+	}
+}
+
+func (m *DesiredRouteManager) UpsertRoute(route DesiredRoute) error {
+	txn := m.db.WriteTxn(m.tbl)
+	defer txn.Abort()
+
+	if err := route.ValidateAndSetDefaults(); err != nil {
+		return err
+	}
+
+	// By default, any new route we add is not selected and does not have to be reconciled.
+	// The [selectRoutes] method will select the best route for each prefix+table.
+	route.selected = false
+	route.SetStatus(reconciler.StatusDone())
+
+	if _, _, err := m.tbl.Insert(txn, &route); err != nil {
+		return err
+	}
+
+	if err := m.selectRoutes(txn, route.GetOwnerlessKey()); err != nil {
+		return err
+	}
+
+	txn.Commit()
+	return nil
+}
+
+func (m *DesiredRouteManager) UpsertRouteWait(route DesiredRoute) error {
+	if err := m.UpsertRoute(route); err != nil {
+		return nil
+	}
+
+	return m.waitForReconciliation(route.GetFullKey())
+}
+
+func (m *DesiredRouteManager) DeleteRoute(route DesiredRoute) error {
+	txn := m.db.WriteTxn(m.tbl)
+	defer txn.Abort()
+
+	if err := route.ValidateAndSetDefaults(); err != nil {
+		return err
+	}
+
+	if _, _, err := m.tbl.Delete(txn, &route); err != nil {
+		return err
+	}
+
+	if err := m.selectRoutes(txn, route.GetOwnerlessKey()); err != nil {
+		return err
+	}
+
+	txn.Commit()
+	return nil
+}
+
+const reconciliationTimeout = 1 * time.Second
+
+func (m *DesiredRouteManager) waitForReconciliation(routeKey DesiredRouteKey) error {
+	t := time.NewTimer(reconciliationTimeout)
+	defer t.Stop()
+
+	var err error
+	for {
+		obj, _, watch, found := m.tbl.GetWatch(m.db.ReadTxn(), DesiredRouteIndex.Query(routeKey))
+		if !found {
+			return fmt.Errorf("route %s not found", routeKey)
+		}
+
+		if obj.status.Kind == reconciler.StatusKindDone {
+			// already reconciled
+			return nil
+		}
+
+		select {
+		case <-t.C:
+			return fmt.Errorf("timeout waiting for parameter %s reconciliation: %w", routeKey, err)
+		case <-watch:
+			if obj.status.Kind == reconciler.StatusKindDone {
+				return nil
+			}
+			if obj.status.Kind == reconciler.StatusKindError {
+				err = errors.New(obj.status.GetError())
+			}
+		}
+	}
+}
+
+func (m *DesiredRouteManager) selectRoutes(txn statedb.WriteTxn, key DesiredRouteKey) error {
+	// Get all routes with the same prefix and table.
+	routes := slices.Collect(statedb.ToSeq(m.tbl.List(txn, DesiredRouteTablePrefixIndex.Query(key))))
+	if len(routes) == 0 {
+		return nil // nothing to select
+	}
+
+	// Sort routes by admin distance and name, so that the first one is the
+	// one that is selected.
+	slices.SortStableFunc(routes, func(a, b *DesiredRoute) int {
+		adminDiff := int(a.AdminDistance) - int(b.AdminDistance)
+		if adminDiff == 0 {
+			if a.Owner.name < b.Owner.name {
+				return -1
+			}
+			if a.Owner.name > b.Owner.name {
+				return 1
+			}
+		}
```

Comment:
I think we want to break the tie of the Priority here.

Reply:
Sorry, never mind. I thought we need a best path selection for the same owner + prefix + AD but with different priority, but such a route can be installed simultaneously to the FIB.

Reply:
> There are valid reasons for having multiple routes with different priority, for example to have a failover route in case the other gets deleted.

Yeah, I mean, if we want to realize this, we should have multiple routes selected if they have an identical keys except priority. No?

Reply:
Yep, I understand that. What I'm saying here is the current selectRouets doesn't seem to implement that semantics.

Reply:
Ah. I did test it, but only with one owner. I think you are correct and the multi-owner scenario does not work currently. I will add a test for that and fix it.